readme.md
# sha256 - calculate SHA-256 message digest

[![GitHub GCC status](https://github.com/striezel/secure-hash-algorithm/workflows/GCC/badge.svg)](https://github.com/striezel/secure-hash-algorithm/actions)
[![GitHub Clang status](https://github.com/striezel/secure-hash-algorithm/workflows/Clang/badge.svg)](https://github.com/striezel/secure-hash-algorithm/actions)
[![GitHub MSYS2 status](https://github.com/striezel/secure-hash-algorithm/workflows/MSYS2/badge.svg)](https://github.com/striezel/secure-hash-algorithm/actions)
[![GitLab pipeline status](https://gitlab.com/striezel/secure-hash-algorithm/badges/master/pipeline.svg)](https://gitlab.com/striezel/secure-hash-algorithm/-/pipelines)

## Purpose of the program

The sha256 program calculates and prints SHA-256 (256bit) checksums
of all files that are passed to it as command line parameters.
It can also be used to calculate SHA-1 (160bit), SHA-224 (224bit),
SHA-384 (384bit) and SHA-512 (512bit) checksums, if desired.

## Program call

Since this (obviously) is a command line program, you can call it
via command line/ shell. A list of valid options follows below.


### Synopsis

  sha256 [--sha1 | --sha224 | --sha256 | --sha384 | --sha512] FILENAME ...


### Options + parameters

  --sha1
      Calculate SHA-1 (160 bit) checksums instead of SHA-256.

  --sha224
      Calculate SHA-224 (224 bit) checksums instead of SHA-256.

  --sha256
      Calculate SHA-256 (256 bit) checksums. This is the default.

  --sha384
      Calculate SHA-384 (384 bit) checksums instead of SHA-256.

  --sha512
      Calculate SHA-512 (512 bit) checksums instead of SHA-256.

  --help
      Show a help message and list valid program options.

  --version
      Print version information and quit.

  FILENAME
        path to a file that should be hashed. Can be repeated
        multiple times.

A typical call could look like:

    sha256 foo.txt some_dir/bar.baz

This would calculate the SHA-256 message digests of foo.txt and
bar.baz in the subdirectory some_dir. If one of the given files does
not exist, the program quits.

## Building from source

### Prerequisites

To build sha256 from source you need a C++ compiler with support for C++11 and
CMake 3.8 or later.
It also helps to have Git, a distributed version control system, on your build
system to get the latest source code directly from the Git repository.

All three can usually be installed be typing

    apt-get install cmake g++ git

or

    yum install cmake gcc-c++ git

into a root terminal.

### Getting the source code

Get the source directly from Git by cloning the Git repository and change to
the directory after the repository is completely cloned:

    git clone https://gitlab.com/striezel/secure-hash-algorithm.git ./sha256
    cd sha256
    git submodule update --init --recursive

The last of the lines above initializes and updates the submodule that the
sha256 source code needs, too, to be build from source.

### Build process

The build process is relatively easy, because CMake does all the preparations.
Starting in the root directory of the source, you can do the following steps:

    mkdir build
    cd build
    cmake ../
    make sha256 -j2

Now the sha256 binary is built and ready for use.

## License, disclaimer and source code

Copyright 2015  Dirk Stolle

The program sha256 is released under the GNU General Public License
version 3, a free software license. For the full text of the license
consult the file LICENSE or view it online at
  <http://www.gnu.org/licenses/gpl-3.0.html>

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHAN-
TABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General
Public License for more details.

The program's source code is published at GitLab.com, the
project itself is located at
  <https://gitlab.com/striezel/secure-hash-algorithm>


ChangeLog.md
# ChangeLog for sha256

## Version 1.5 (2022-11-12)
  - An error in the program's help text is fixed.
  - The minimum required CMake version to build the program is raised to
    CMake 3.8.

## Version 1.4 (2015-08-06)
  - add option to use SHA-384 instead of SHA-256

## Version 1.3 (2015-08-03)
  - add option to hash with SHA-512 instead of SHA-256

## Version 1.2 (2015-08-01)
  - add option to hash with SHA-224 instead of SHA-256

## Version 1.1 (2012-09-18)
  - add option to hash with SHA-1 instead of SHA-256

## Version 1.0 (2012-08-30)
  - initial version


tests/sha256/additional/main.cpp
/*
 -----------------------------------------------------------------------------
    This file is part of a test suite for a secure hashing algorithm program.
    Copyright (C) 2015  Thoronador

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 -----------------------------------------------------------------------------
*/

#include <fstream>
#include <iostream>
#include <string>
#include <tuple>
#include <vector>
#include "../../../libthoro/filesystem/DirectoryFunctions.hpp"
#include "../../../libthoro/filesystem/FileFunctions.hpp"
#include "../../../libthoro/hash/sha256/sha256.hpp"
#include "../../../libthoro/hash/sha256/FileSourceUtility.hpp"

/*
  The following example test messages and digests are take from
  <http://csrc.nist.gov/groups/ST/toolkit/documents/Examples/SHA2_Additional.pdf>.
*/

const std::vector<std::tuple<std::size_t, char, std::string> > testData =
{
  // #3
    std::tuple<std::size_t, char, std::string>(55, '\0',
        "02779466cdec163811d078815c633f21901413081449002f24aa3e80f0b88ef7"),
    // #4
    std::tuple<std::size_t, char, std::string>(56, '\0',
        "d4817aa5497628e7c77e6b606107042bbba3130888c5f47a375e6179be789fbb"),
    // #5
    std::tuple<std::size_t, char, std::string>(57, '\0',
        "65a16cb7861335d5ace3c60718b5052e44660726da4cd13bb745381b235a1785"),
    // #6
    std::tuple<std::size_t, char, std::string>(64, '\0',
        "f5a5fd42d16a20302798ef6ed309979b43003d2320d9f0e8ea9831a92759fb4b"),
    // #7
    std::tuple<std::size_t, char, std::string>(1000, '\0',
        "541b3e9daa09b20bf85fa273e5cbd3e80185aa4ec298e765db87742b70138a53"),
    // #8
    std::tuple<std::size_t, char, std::string>(1000, 0x41 /* A */,
        "c2e686823489ced2017f6059b8b239318b6364f6dcd835d0a519105a1eadd6e4"),
    // #9
    std::tuple<std::size_t, char, std::string>(1005, 0x55 /* U */,
        "f4d62ddec0f3dd90ea1380fa16a5ff8dc4c54b21740650f24afc4120903552b0"),
    // #10
    std::tuple<std::size_t, char, std::string>(1000000, '\0',
        "d29751f2649b32ff572b5e0a9f541ea660a50f94ff0beedfb0b692b924cc8025"),

    // #11
    std::tuple<std::size_t, char, std::string>(536870912, 0x5a /* 'Z' */,
        "15a1868c12cc53951e182344277447cd0979536badcc512ad24c67e9b2d4f3dd"),
    // #12
    std::tuple<std::size_t, char, std::string>(1090519040, '\0',
        "461c19a93bd4344f9215f5ec64357090342bc66b15a148317d276e31cbc20b53"),
    // #13
    std::tuple<std::size_t, char, std::string>(1610612798, 0x42 /* 'B' */,
        "c23ce8a7895f4b21ec0daf37920ac0a262a220045a03eb2dfed48ef9b05aabea")
};

// maximum file size for "short" file tests
const std::size_t cShortMaximum = 1024*1024;

int main(int argc, char** argv)
{
  if (argc<2)
  {
    std::cout << "Not enough command line arguments!" << std::endl;
    return 1;
  }

  //directory for sample files
  std::string fileDir = "";
  //switch for hashing longer files (default: false)
  bool hashLongFiles = false;

  if (argc>1)
  {
    if (argv!=nullptr)
    {
      //first parameter: directory with first test files
      if (argv[1] != nullptr)
      {
        fileDir = std::string(argv[1]);
      } //if argv[1] is set
      //second parameter: allow long files
      if (argv[2] != nullptr)
      {
        const std::string param(argv[2]);
        if (param=="--long" || param=="-l")
        {
          hashLongFiles = true;
        } //if
        else
        {
          std::cout << "Error: Invalid parameter " << param << "!" << std::endl;
          return 1;
        }
      } //if argv[2] exists

    } //if argv != null
  } //if args are present

  if (!libthoro::filesystem::Directory::exists(fileDir))
  {
    std::cout << "Error: Directory " << fileDir << " does not exist!"
              << std::endl;
    return 1;
  }
  //add trailing slash, if it is missing
  fileDir = libthoro::filesystem::slashify(fileDir);

  if (!hashLongFiles)
  {
    std::cout << "Info: Skipping test data with messages longer than "
              << cShortMaximum << " bytes." << std::endl;
  }
  else
  {
    std::cout << "Info: Processing test data with messages longer than "
              << cShortMaximum << " bytes, too." << std::endl;
  }

  //hash first file
  SHA256::MessageDigest md_sha256 = SHA256::computeFromFile(fileDir+"file_1_byte_0xbd.dat");
  std::cout << "File: file_1_byte_0xbd.dat" << std::endl
            << "Expected digest:   68325720aabd7c82f30f554b313d0570c95accbb7dc4b5aae11204c08ffe732b" << std::endl
            << "Calculated digest: " << md_sha256.toHexString() << std::endl;
  if ("68325720aabd7c82f30f554b313d0570c95accbb7dc4b5aae11204c08ffe732b" != md_sha256.toHexString())
  {
    std::cout << "ERROR: Message digest is not as expected!" << std::endl;
    return 1;
  }
  //hash second file
  md_sha256 = SHA256::computeFromFile(fileDir+"file_4_bytes_0xc98c8e55.dat");
  std::cout << "File: file_4_bytes_0xc98c8e55.dat" << std::endl
            << "Expected digest:   7abc22c0ae5af26ce93dbb94433a0e0b2e119d014f8e7f65bd56c61ccccd9504" << std::endl
            << "Calculated digest: " << md_sha256.toHexString() << std::endl;
  if ("7abc22c0ae5af26ce93dbb94433a0e0b2e119d014f8e7f65bd56c61ccccd9504" != md_sha256.toHexString())
  {
    std::cout << "ERROR: Message digest is not as expected!" << std::endl;
    return 1;
  }

  //create temp file for tests
  std::string fileName = "";
  if (!libthoro::filesystem::File::createTemp(fileName))
  {
    std::cout << "Error: Could not create temporary file!"   << std::endl;
    return 1;
  }

  for (auto && item : testData)
  {
    if (hashLongFiles || (std::get<0>(item)<= cShortMaximum))
    {
      //generate file from data
      std::ofstream stream;
      stream.open(fileName, std::ios_base::trunc | std::ios_base::binary | std::ios_base::out);
      if (!stream.good() || !stream.is_open())
      {
        std::cout << "Error: could not create/open temporary file!" << std::endl;
        return 1;
      }
      std::size_t i = 0;
      for (i=0; i<std::get<0>(item); ++i)
      {
        stream.put(std::get<1>(item));
      } //for
      if (!stream.good())
      {
        std::cout << "Error: Could not write data to temporary file!" << std::endl;
        stream.close();
        libthoro::filesystem::File::remove(fileName);
        return 1;
      }
      stream.close();

      //reset hash
      md_sha256.setToNull();
      md_sha256 = SHA256::computeFromFile(fileName);
      std::cout << "Message: "<< std::get<0>(item) << " bytes of " << static_cast<unsigned int>(std::get<1>(item)) << std::endl
                << "Expected digest:   " << std::get<2>(item) << std::endl
                << "Calculated digest: " << md_sha256.toHexString() << std::endl;
      if (std::get<2>(item) != md_sha256.toHexString())
      {
        std::cout << "ERROR: Message digest is not as expected!" << std::endl;
        libthoro::filesystem::File::remove(fileName);
        return 1;
      }
    } //if short/long check
  } //for
  libthoro::filesystem::File::remove(fileName);
  std::cout << "Passed test!" << std::endl;
  return 0;
}


tests/sha224/additional-file/main.cpp
/*
 -----------------------------------------------------------------------------
    This file is part of a test suite for a secure hashing algorithm program.
    Copyright (C) 2015, 2016  Dirk Stolle

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 -----------------------------------------------------------------------------
*/

#include <fstream>
#include <iostream>
#include <string>
#include <tuple>
#include <vector>
#include "../../../libstriezel/filesystem/directory.hpp"
#include "../../../libstriezel/filesystem/file.hpp"
#include "../../../libstriezel/hash/sha224/sha224.hpp"
#include "../../../libstriezel/hash/sha224/FileSourceUtility.hpp"

/*
  The following example test messages and digests are take from
  <http://csrc.nist.gov/groups/ST/toolkit/documents/Examples/SHA2_Additional.pdf>.
*/

const std::vector<std::tuple<std::size_t, char, std::string> > testData =
{
    // #3
    std::tuple<std::size_t, char, std::string>(56, '\0',
        "5c3e25b69d0ea26f260cfae87e23759e1eca9d1ecc9fbf3c62266804"),
    // #4
    std::tuple<std::size_t, char, std::string>(1000, 0x51, /* Q */
        "3706197f66890a41779dc8791670522e136fafa24874685715bd0a8a"),
    // #5
    std::tuple<std::size_t, char, std::string>(1000, 0x41, /* A */
        "a8d0c66b5c6fdfd836eb3c6d04d32dfe66c3b1f168b488bf4c9c66ce"),
    // #6
    std::tuple<std::size_t, char, std::string>(1005, 0x99,
        "cb00ecd03788bf6c0908401e0eb053ac61f35e7e20a2cfd7bd96d640"),
    // #7
    std::tuple<std::size_t, char, std::string>(1000000, '\0',
        "3a5d74b68f14f3a4b2be9289b8d370672d0b3d2f53bc303c59032df3"),
    // #8
    std::tuple<std::size_t, char, std::string>(536870912, 0x41 /* A */,
        "c4250083cf8230bf21065b3014baaaf9f76fecefc21f91cf237dedc9"),
    // #9
    std::tuple<std::size_t, char, std::string>(1090519040, '\0',
        "014674abc5cb980199935695af22fab683748f4261d4c6492b77c543"),
    // #10
    std::tuple<std::size_t, char, std::string>(1610612799, 0x84,
        "a654b50b767a8323c5b519f467d8669837142881dc7ad368a7d5ef8f")
};

// maximum file size for "short" file tests
const std::size_t cShortMaximum = 1024*1024;

int main(int argc, char** argv)
{
  if (argc<2)
  {
    std::cout << "Not enough command line arguments!" << std::endl;
    return 1;
  }

  //directory for sample files
  std::string fileDir = "";
  //switch for hashing longer files (default: false)
  bool hashLongFiles = false;

  if (argc>1)
  {
    if (argv!=nullptr)
    {
      //first parameter: directory with first test files
      if (argv[1] != nullptr)
      {
        fileDir = std::string(argv[1]);
      } //if argv[1] is set
      //second parameter: allow long files
      if (argv[2] != nullptr)
      {
        const std::string param(argv[2]);
        if (param=="--long" || param=="-l")
        {
          hashLongFiles = true;
        } //if
        else
        {
          std::cout << "Error: Invalid parameter " << param << "!" << std::endl;
          return 1;
        }
      } //if argv[2] exists

    } //if argv != null
  } //if args are present

  if (!libstriezel::filesystem::directory::exists(fileDir))
  {
    std::cout << "Error: Directory " << fileDir << " does not exist!"
              << std::endl;
    return 1;
  }
  //add trailing slash, if it is missing
  fileDir = libstriezel::filesystem::slashify(fileDir);

  if (!hashLongFiles)
  {
    std::cout << "Info: Skipping test data with messages longer than "
              << cShortMaximum << " bytes." << std::endl;
  }
  else
  {
    std::cout << "Info: Processing test data with messages longer than "
              << cShortMaximum << " bytes, too." << std::endl;
  }

  //hash first file
  SHA224::MessageDigest md_sha224 = SHA224::computeFromFile(fileDir+"file_1_byte_0xff.dat");
  std::cout << "File: file_1_byte_0xff.dat" << std::endl
            << "Expected digest:   e33f9d75e6ae1369dbabf81b96b4591ae46bba30b591a6b6c62542b5" << std::endl
            << "Calculated digest: " << md_sha224.toHexString() << std::endl;
  if ("e33f9d75e6ae1369dbabf81b96b4591ae46bba30b591a6b6c62542b5" != md_sha224.toHexString())
  {
    std::cout << "ERROR: Message digest is not as expected!" << std::endl;
    return 1;
  }
  //hash second file
  md_sha224 = SHA224::computeFromFile(fileDir+"file_4_bytes_0xe5e09924.dat");
  std::cout << "File: file_4_bytes_0xe5e09924.dat" << std::endl
            << "Expected digest:   fd19e74690d291467ce59f077df311638f1c3a46e510d0e49a67062d" << std::endl
            << "Calculated digest: " << md_sha224.toHexString() << std::endl;
  if ("fd19e74690d291467ce59f077df311638f1c3a46e510d0e49a67062d" != md_sha224.toHexString())
  {
    std::cout << "ERROR: Message digest is not as expected!" << std::endl;
    return 1;
  }

  //create temp file for tests
  std::string fileName = "";
  if (!libstriezel::filesystem::file::createTemp(fileName))
  {
    std::cout << "Error: Could not create temporary file!"   << std::endl;
    return 1;
  }

  for (auto && item : testData)
  {
    if (hashLongFiles || (std::get<0>(item)<= cShortMaximum))
    {
      //generate file from data
      std::ofstream stream;
      stream.open(fileName, std::ios_base::trunc | std::ios_base::binary | std::ios_base::out);
      if (!stream.good() || !stream.is_open())
      {
        std::cout << "Error: could not create/open temporary file!" << std::endl;
        return 1;
      }
      std::size_t i = 0;
      for (i=0; i<std::get<0>(item); ++i)
      {
        stream.put(std::get<1>(item));
      } //for
      if (!stream.good())
      {
        std::cout << "Error: Could not write data to temporary file!" << std::endl;
        stream.close();
        libstriezel::filesystem::file::remove(fileName);
        return 1;
      }
      stream.close();

      //reset hash
      md_sha224.setToNull();
      md_sha224 = SHA224::computeFromFile(fileName);
      std::cout << "Message: "<< std::get<0>(item) << " bytes of " << static_cast<unsigned int>(std::get<1>(item)) << std::endl
                << "Expected digest:   " << std::get<2>(item) << std::endl
                << "Calculated digest: " << md_sha224.toHexString() << std::endl;
      if (std::get<2>(item) != md_sha224.toHexString())
      {
        std::cout << "ERROR: Message digest is not as expected!" << std::endl;
        libstriezel::filesystem::file::remove(fileName);
        return 1;
      }
    } //if short/long check
  } //for
  libstriezel::filesystem::file::remove(fileName);
  std::cout << "Passed test!" << std::endl;
  return 0;
}


ci/cpplint.sh
#!/bin/bash

#  cpplint.sh - script to check the syntax of C++ source files
#
#  Copyright (C) 2015  Dirk Stolle
#
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>.

if [[ -z $CXX ]]
then
  # no C++ compiler is set, assume GNU C++ compiler
  CXX=g++
fi

# CFLAGS need to be set for some files in libthoro/, because they require the
# include path for libxml2 headers.
if [[ -x /usr/bin/xml2-config ]]
then
  CFLAGS=$(/usr/bin/xml2-config --cflags)
else
  # just use some standard path for libxml2 includes
  CFLAGS="-I/usr/include/libxml2"
fi

# print $CXX version to see which version is used for syntax check
$CXX --version
echo

# find all .sh files and run them through Bash's syntax check
find ./ \( -name '*.cpp' -o -name '*.hpp' -o -name '*.h' \) -print0 | xargs -0 -i $CXX $CFLAGS -std=c++0x -fsyntax-only -Wall {}
if [[ $? -ne 0 ]]
then
  echo "Some source code files contain syntax errors!"
  echo "You should do something about it."
  echo 'And do it "soon(TM)".'
  exit 1
else
  echo "Syntax seems to be correct."
  echo "Please take this happy smilie with you.  :)"
  exit 0
fi


sha256/main.cpp
/*
 -------------------------------------------------------------------------------
    This file is part of the SHA-256 hash calculator.
    Copyright (C) 2012, 2015, 2016, 2022  Dirk Stolle

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 -------------------------------------------------------------------------------
*/

#include <iostream>
#include <set>
#include <string>
#include "../libstriezel/filesystem/file.hpp"
#include "../libstriezel/hash/sha512/sha512.hpp"
#include "../libstriezel/hash/sha512/FileSource.hpp"
#include "../libstriezel/hash/sha512/FileSourceUtility.hpp"
#include "../libstriezel/hash/sha384/sha384.hpp"
#include "../libstriezel/hash/sha384/FileSource.hpp"
#include "../libstriezel/hash/sha384/FileSourceUtility.hpp"
#include "../libstriezel/hash/sha256/sha256.hpp"
#include "../libstriezel/hash/sha256/FileSource.hpp"
#include "../libstriezel/hash/sha256/FileSourceUtility.hpp"
#include "../libstriezel/hash/sha224/sha224.hpp"
#include "../libstriezel/hash/sha224/FileSource.hpp"
#include "../libstriezel/hash/sha224/FileSourceUtility.hpp"
#include "../libstriezel/hash/sha1/sha1.hpp"
#include "../libstriezel/hash/sha1/FileSource.hpp"
#include "../libstriezel/hash/sha1/FileSourceUtility.hpp"

//return codes
const int rcInvalidParameter = 1;

void showGPLNotice()
{
  std::cout << "SHA-256 file hash calculator\n"
            << "  Copyright (C) 2012, 2015, 2022  Dirk Stolle\n"
            << "\n"
            << "  This program is free software: you can redistribute it and/or\n"
            << "  modify it under the terms of the GNU General Public License as published\n"
            << "  by the Free Software Foundation, either version 3 of the License, or\n"
            << "  (at your option) any later version.\n"
            << "\n"
            << "  This program is distributed in the hope that they will be useful,\n"
            << "  but WITHOUT ANY WARRANTY; without even the implied warranty of\n"
            << "  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the\n"
            << "  GNU General Public License for more details.\n"
            << "\n"
            << "  You should have received a copy of the GNU General Public License\n"
            << "  along with this program.  If not, see <http://www.gnu.org/licenses/>.\n"
            << "\n";
}

void showVersion()
{
  showGPLNotice();
  std::cout << "SHA-256 file hash calculator, version 1.5, 2022-11-12\n";
}

void showHelp()
{
  std::cout << "\nsha256 [--sha1 | --sha224 | --sha256 | --sha384 | --sha512] FILENAME\n"
            << "\n"
            << "options:\n"
            << "  --help           - Displays this help message and quits.\n"
            << "  -?               - same as --help\n"
            << "  --version        - Displays the version of the program and quits.\n"
            << "  -v               - same as --version\n"
            << "  FILENAME         - Set path to file that should be hashed. Can be repeated\n"
            << "                     multiple times. Has to appear at least once.\n"
            << "  --sha1           - Use SHA-1 instead of SHA-256 to hash files.\n"
            << "  --sha224         - Use SHA-224 instead of SHA-256 to hash files.\n"
            << "  --sha256         - Use SHA-256 to hash files. This option is active by\n"
            << "                     default.\n"
            << "  --sha384         - Use SHA-384 instead of SHA-256 to hash files.\n"
            << "  --sha512         - Use SHA-512 instead of SHA-256 to hash files.\n";
}

int main(int argc, char **argv)
{
  std::set<std::string> files;

  enum SHAHashType { htUnspecified, htSHA1, htSHA224, htSHA256, htSHA384, htSHA512 };

  SHAHashType hashType = htUnspecified;

  if ((argc > 1) && (argv != nullptr))
  {
    int i = 1;
    while (i < argc)
    {
      if (argv[i] != nullptr)
      {
        const std::string param = std::string(argv[i]);
        // help parameter
        if ((param == "--help") || (param == "-?") || (param == "/?"))
        {
          showHelp();
          return 0;
        }
        // version information requested?
        else if ((param == "--version") || (param == "-v"))
        {
          showVersion();
          return 0;
        }
        else if ((param == "--sha1") || (param == "--sha-1") || (param == "--sha160")
              || (param == "--sha-160"))
        {
          if (hashType == htSHA1)
          {
            std::cerr << "Error: Parameter " << param << " must not occur more than once!\n";
            return rcInvalidParameter;
          }
          if (hashType != htUnspecified)
          {
            std::cerr << "Error: Parameter " << param << " must not occur "
                      << "after hash type has already been set!\n";
            return rcInvalidParameter;
          }
          hashType = htSHA1;
        } // sha-1
        else if ((param == "--sha224") || (param == "--sha-224"))
        {
          if (hashType == htSHA224)
          {
            std::cerr << "Error: Parameter " << param << " must not occur more than once!\n";
            return rcInvalidParameter;
          }
          if (hashType != htUnspecified)
          {
            std::cerr << "Error: Parameter " << param << " must not occur "
                      << "after hash type has already been set!\n";
            return rcInvalidParameter;
          }
          hashType = htSHA224;
        } // sha-224
        else if ((param == "--sha256") || (param == "--sha-256"))
        {
          if (hashType == htSHA256)
          {
            std::cerr << "Error: Parameter " << param << " must not occur more than once!\n";
            return rcInvalidParameter;
          }
          if (hashType != htUnspecified)
          {
            std::cerr << "Error: Parameter " << param << " must not occur "
                      << "after hash type has already been set!\n";
            return rcInvalidParameter;
          }
          hashType = htSHA256;
        } // sha-256
        else if ((param == "--sha384") || (param == "--sha-384"))
        {
          if (hashType == htSHA384)
          {
            std::cerr << "Error: Parameter " << param << " must not occur more than once!\n";
            return rcInvalidParameter;
          }
          if (hashType != htUnspecified)
          {
            std::cerr << "Error: Parameter " << param << " must not occur "
                      << "after hash type has already been set!\n";
            return rcInvalidParameter;
          }
          hashType = htSHA384;
        } // sha-384
        else if ((param == "--sha512") || (param == "--sha-512"))
        {
          if (hashType == htSHA512)
          {
            std::cerr << "Error: Parameter " << param << " must not occur more than once!\n";
            return rcInvalidParameter;
          }
          if (hashType != htUnspecified)
          {
            std::cerr << "Error: Parameter " << param << " must not occur "
                      << "after hash type has already been set!\n";
            return rcInvalidParameter;
          }
          hashType = htSHA512;
        } // sha-512
        else
        {
          // should be filename
          if (libstriezel::filesystem::file::exists(param))
          {
            // add file to list
            files.insert(param);
          }
          else
          {
            std::cerr << "Invalid parameter/filename given: \"" << param
                      << "\" does not name an existing file!\n"
                      << "Use --help to get a list of valid parameters.\n";
            return rcInvalidParameter;
          }
        }
      } // parameter exists
      else
      {
        std::cerr << "Parameter at index " << i << "  is NULL.\n";
        return rcInvalidParameter;
      }
      ++i; // on to next parameter
    }
  }
  else
  {
    std::cerr << "You have to specify certain parameters for this program to run properly.\n"
              << "Use --help to get a list of valid parameters.\n";
    return rcInvalidParameter;
  }

  if (files.empty())
  {
    std::cerr << "You have to specify certain parameters for this program to run properly.\n"
              << "Use --help to get a list of valid parameters.\n";
    return rcInvalidParameter;
  }

  // Set default hash algorithm, if no choice was made.
  if (hashType == htUnspecified)
    hashType = htSHA256;

  std::cout << "Hashing file(s), this may take a while..." << std::endl;

  SHA512::MessageDigest hash512;
  SHA384::MessageDigest hash384;
  SHA256::MessageDigest hash256;
  SHA224::MessageDigest hash224;
  SHA1::MessageDigest hash160;
  for (const auto& item: files)
  {
    switch (hashType)
    {
      case htSHA1:
           hash160 = SHA1::computeFromFile(item);
           std::cout << hash160.toHexString() << "  " << item << std::endl;
           break;
      case htSHA224:
           hash224 = SHA224::computeFromFile(item);
           std::cout << hash224.toHexString() << "  " << item << std::endl;
           break;
      case htSHA384:
           hash384 = SHA384::computeFromFile(item);
           std::cout << hash384.toHexString() << "  " << item << std::endl;
           break;
      case htSHA512:
           hash512 = SHA512::computeFromFile(item);
           std::cout << hash512.toHexString() << "  " << item << std::endl;
           break;
      default:
           hash256 = SHA256::computeFromFile(item);
           std::cout << hash256.toHexString() << "  " << item << std::endl;
           break;
    }
  }

  return 0;
}
